### PR TITLE
Adding telemetry command extension to Agent

### PR DESCRIPTION
**Pull request by nigurr**

### CODE REVIEW

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                startOfSegment = index;
+            }
+            op.Add(input.Substring(startOfSegment).Replace(escapeCharacter + separator, separator));
+
+            return op;
+        }
+
+        internal static class WellKnownEventTrackCommand
+        {
+            internal static readonly string TrackEvent = "trackevent";
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: maybe just call ##vso[telemetry.publish]?

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+            // Hook up Throttling event, we will log warning on server tarpit.
+            ciService.CustomerIntelligenceQueueThrottling += ServiceThrottling_EventReceived;
+            if (_throttlingReported)
+            {
+                context.Warning(StringUtil.Loc("ServerTarpit"));
+            }
+        }
+
+        private void LoadTelemetryInputs(Dictionary<string, string> eventProperties)
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: just put code inline, since the method only used once.

File: src/Agent.Worker/Telemetry/CustomerIntelligenceService.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Telemetry
+{
+    [ServiceLocator(Default = typeof(CustomerIntelligenceService))]
+    public interface ICustomerIntelligenceService : IAgentService, IThrottlingReporter
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/CustomerIntelligenceService.cs]: in the agent project, we normally call this `ICustomerIntelligenceServer `

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            var ciService = HostContext.GetService<ICustomerIntelligenceService>();
+            var vssConnection = WorkerUtilies.GetVssConnection(context, new DelegatingHandler[] { new ThrottlingReportHandler(ciService) });
+            ciService.Initialize(vssConnection);
+
+            var commandContext = HostContext.CreateService<IAsyncCommandContext>();
+            commandContext.InitializeCommandContext(context, StringUtil.Loc("Telemetry"));
+            commandContext.Task = ciService.PublishEventsAsync(new CustomerIntelligenceEvent[] { PopulateCustomerIntelligenceData(context, data) });
+            context.AsyncCommands.Add(commandContext);
+
+            // Hook up Throttling event, we will log warning on server tarpit.
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: we shouldn't just log warning, we need backoff and retry.

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            ciService.CustomerIntelligenceQueueThrottling += ServiceThrottling_EventReceived;
+            if (_throttlingReported)
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: will this work? you check the flag in may context, while the async command is running.

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs (hunk 1 above)
Comment (nigurr) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: Will fix this

File: src/Agent.Worker/Telemetry/CustomerIntelligenceService.cs (hunk 3 above)
Comment (nigurr) [inline src/Agent.Worker/Telemetry/CustomerIntelligenceService.cs]: will fix this

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs (hunk 4 above)
Comment (nigurr) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: I removed throttling for now as it is common for any task command not specific to CI. I will take it as separate PR as it requires some change in design.

File: src/Agent.Worker/Telemetry/CustomerIntelligenceServer.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        void Initialize(VssConnection connection);
+        Task PublishEventsAsync(CustomerIntelligenceEvent[] ciEvents);
+    }
+
+    // This service is used for tracking task events which are applicable for VSTS internal tasks
+    public class CustomerIntelligenceServer : AgentService, ICustomerIntelligenceServer
+    {
+        private CustomerIntelligenceHttpClient _ciClient;
+
+        public void Initialize(VssConnection connection)
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/CustomerIntelligenceServer.cs]: copy the pattern other server use. Ex: agentserver.cs
```
private bool _hasConnection;
private VssConnection _connection;
private TaskHttpClient _taskClient;
```

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            var vssConnection = WorkerUtilies.GetVssConnection(context);
+            ciService.Initialize(vssConnection);
+
+            var commandContext = HostContext.CreateService<IAsyncCommandContext>();
+            commandContext.InitializeCommandContext(context, StringUtil.Loc("Telemetry"));
+            commandContext.Task = ciService.PublishEventsAsync(new CustomerIntelligenceEvent[] { PopulateCustomerIntelligenceData(context, data) });
+            context.AsyncCommands.Add(commandContext);
+        }
+
+        private void LoadTelemetryInputs(Dictionary<string, string> eventProperties)
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: do this inline, no need for create a method only get called once.

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        private CustomerIntelligenceEvent PopulateCustomerIntelligenceData(IExecutionContext context, string data)
+        {
+            if (string.IsNullOrEmpty(data))
+            {
+                throw new ArgumentException(StringUtil.Loc("ArgumentNeeded", "EventTrackerData"));
+            }
+
+            // Expected format is key1=Value1;key2=Value2;Key3=Value3
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: can we just use json? since you just need to deserialize a string into dictionary, i want to invent a new format.
{a:b,c:d,...}

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+            LoadTelemetryInputs(eventProperties);
+
+            var ciService = HostContext.GetService<ICustomerIntelligenceServer>();
+            var vssConnection = WorkerUtilies.GetVssConnection(context);
+            ciService.Initialize(vssConnection);
+
+            var commandContext = HostContext.CreateService<IAsyncCommandContext>();
+            commandContext.InitializeCommandContext(context, StringUtil.Loc("Telemetry"));
+            commandContext.Task = ciService.PublishEventsAsync(new CustomerIntelligenceEvent[] { PopulateCustomerIntelligenceData(context, data) });
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: have you try the behavior when you get tarpit?
https://vsowiki.com/index.php?title=Resource_Utilization_Service

File: src/Agent.Worker/WorkerUtilties.cs [hunk 10]
```diff
@@ -1,15 +1,17 @@
 ﻿using Microsoft.TeamFoundation.DistributedTask.WebApi;
```
Comment (TingluoHuang) [inline src/Agent.Worker/WorkerUtilties.cs]: undo the change since you didn't use them

File: src/Agent.Worker/Telemetry/CustomerIntelligenceServer.cs (hunk 6 above)
Comment (nigurr) [inline src/Agent.Worker/Telemetry/CustomerIntelligenceServer.cs]: We are not reusing `_connection` property. and almost all Server classes like 'BuildServer' are using getClient directly as vssConnection is already made in WorkerUtil.GetConnection

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs (hunk 8 above)
Comment (nigurr) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: Thought about it. since mostly our commands are following this pattern `key=value;`, I tried to maintain the same here. 
We can also serialize and deserialize however our commands dont support multi-line. so one has to serialize into compact mode and doesnt sound great to me.
telemetry is something to track minute data points and its not a logger where someone should send big objects, data.

**Comment by nigurr** [comment]
@TingluoHuang if there are not any major concerns, can merge the PR.

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs (hunk 9 above)
Comment (nigurr) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: @TingluoHuang yes. since it's common across any TFS service/and any command that's using TFS service api essentially needs to handle this.
It's best to give such functionality overall for AsyncCommandContext instead of something specific to Customer Intelligence, which we take it as separate PR.

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Telemetry
+{
+    public class TelemetryCommandExtension : AgentService, IWorkerCommandExtension
+    {
+        private string _area;
+        private string _feature;
+        public HostTypes SupportedHostTypes => HostTypes.Build;
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: only for build? what about RM?

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs (hunk 8 above)
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: what your command looks like?
here is what i thought it should be:
```
##vso[telemetry.publish area=test;feature=somefeature;]{foo:bar...}
```
Since all you need get a Dictionary<String, String> from the command payload, i feel JSON makes prefect sense.
and we don't have to worry about escaping problem at all.

Again, I am not interested in invent a new format which might introduce new escape problem.

File: src/Agent.Worker/WorkerUtilties.cs (hunk 10 above)
Comment (TingluoHuang) [inline src/Agent.Worker/WorkerUtilties.cs]: you still have changes to this file.

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs (hunk 7 above)
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: do this inline.

File: src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Telemetry
+{
+    public class TelemetryCommandExtension : AgentService, IWorkerCommandExtension
+    {
+        private string _area;
```
Comment (TingluoHuang) [inline src/Agent.Worker/Telemetry/TelemetryCommandExtension.cs]: not need a global var _area/_feature, do everything inline.

**Comment by nigurr** [comment]
@TingluoHuang 
1. Area/Command are Class variables like other extensions. https://github.com/Microsoft/vsts-agent/blob/master/src/Agent.Worker/TestResults/ResultsCommandExtension.cs#L29
I believe we invoke each instance per each command so this shouldn't be a problem. Correct me please.

2. LoadTelemetryInputs - do inline - it's nit and it's good to have well defined short methods

3. JSOn - I addressed it. we are no longer using `;` separated data. We are using 'JSON' data.
